src/multiclient/multi_client_actor.cpp

```diff
@@ -0,0 +1,266 @@
+  int32_t last_mc_seqno_value = last_mc_seqno.value_or(-1);
+
+  LOG(DEBUG) << "LS #" << worker_index << " is_alive: " << is_alive << " last_mc_seqno: " << last_mc_seqno_value;
+
+  auto& worker = workers_[worker_index];
+  worker.is_alive = is_alive;
+  worker.is_waiting_for_update = false;
+
+  if (is_alive) {
+    worker.last_mc_seqno = last_mc_seqno_value;
+    worker.check_retry_count = 0;
+  } else {
+    worker.check_retry_after = td::Timestamp::in(kRetryInterval);
+  }
+}
+
+void MultiClientActor::check_archival() {
+  static constexpr int32_t kBlockWorkchain = ton::masterchainId;
+  static constexpr int64_t kBlockShard = ton::shardIdAll;
+  static constexpr int32_t kBlockSeqno = 3;
+
+  static constexpr int kLookupMode = 1;
+  static constexpr int kLookupLt = 0;
+  static constexpr int kLookupUtime = 0;
+
+  for (size_t worker_index = 0; worker_index < workers_.size(); worker_index++) {
+    if (!workers_[worker_index].is_alive) {
+      continue;
+    }
+
+    send_worker_request<ton::tonlib_api::blocks_lookupBlock>(
+        worker_index,
+        ton::tonlib_api::blocks_lookupBlock(
+            kLookupMode,
+            ton::tonlib_api::make_object<ton::tonlib_api::ton_blockId>(kBlockWorkchain, kBlockShard, kBlockSeqno),
+            kLookupLt,
+            kLookupUtime
+        ),
+        [self_id = actor_id(this), worker_index](auto result) {
+          td::actor::send_closure(self_id, &MultiClientActor::on_archival_checked, worker_index, result.is_ok());
+        }
+    );
+  }
+}
+
+void MultiClientActor::on_archival_checked(size_t worker_index, bool is_archival) {
+  LOG(DEBUG) << "LS #" << worker_index << " archival: " << is_archival;
+  workers_[worker_index].is_archival = is_archival;
+}
+
+std::vector<size_t> MultiClientActor::select_workers(const RequestParameters& options) const {
+  std::vector<size_t> result;
+  result.reserve(workers_.size());
+
+  for (size_t i : std::views::iota(0u, workers_.size()) |
+           std::views::filter([&](size_t i) { return workers_[i].is_alive; }) |
+           std::views::filter([&](size_t i) { return options.archival == true ? workers_[i].is_archival : true; })) {
+    result.push_back(i);
+  }
+
+  if (result.empty()) {
+    return result;
+  }
+
+  switch (options.mode) {
+    case RequestMode::Broadcast:
+      return result;
+
+    case RequestMode::Single: {
+      if (options.lite_server_indexes.has_value()) {
+        CHECK(!options.lite_server_indexes->empty() && options.lite_server_indexes->size() == 1);
+
+        return std::find(result.begin(), result.end(), options.lite_server_indexes->front()) != result.end() ?
+            std::vector<size_t>{options.lite_server_indexes.value().front()} :
+            std::vector<size_t>{};
+      }
+
+      return std::vector<size_t>{result[get_random_index<size_t>(0, result.size() - 1)]};
+    }
+
+    case RequestMode::Multiple: {
+      CHECK(!(options.clients_number.has_value() && options.lite_server_indexes.has_value()));
```

Comment:
maybe these checks would better fit in `RequestParameters`

Reply:
I thought about that, but there is no easy way to do that. 
or you mean make method like: `.validate` for RequestParameters struct ?

Reply:
Yes, `.validate` sounds good. Also, I think CHECK is not the best macro for validating user input since it produces a fatal error, ideally it should provide error back to the user without crashing.